compose/foundation/foundation/src/jsWasmMain/kotlin/androidx/compose/foundation/v2/Actuals.jsWasm.kt

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2023 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.compose.foundation.v2
+
+import kotlinx.coroutines.*
+
+
+private val mainScope = MainScope()
+
+internal actual fun runBlockingCoroutine(block: suspend CoroutineScope.() -> Unit) {
+    // This launch should return only after block is finished
```

Comment:
How does it work on JS where we don't have ability to run something blocking?

What happens, if ScrollableState has suspensions points in `suspend fun scroll` function?

Reply:
You're right. The comment lies. 

> // This launch should return only after block is finished

To test this use case I run our example1 on desktop with:

```kotlin
    private var rawPosition: Double
        get() = scrollScale * adapter.scrollOffset
        set(value) {
            GlobalScope.launch(start = CoroutineStart.UNDISPATCHED) {
                adapter.scrollTo(value / scrollScale)
            }
//            runBlockingCoroutine {
//                adapter.scrollTo(value / scrollScale)
//            }
        }
```
When I run it on the example it looks fine. The desktop tests also pass. 

But it works now only thanks to implementation details, which can change in the future. Adding some `yield()` in the implementation breaks the tests.


> How does it work on JS where we don't have ability to run something blocking?
The current implementation of scroll finishes completely without suspensions. But runBlockingCoroutine is not blocking if something suspends.
____

Ideally, we would make runBlocking unnecessary here to have the same implementation everywhere. I'll try to think if it's possible to tweak the implementation. @m-sasha Sasha, what do you think?

Otherwise, the web implementation will be different.

Reply:
I have a draft of asynchronous implementation. I'll probably share it soon as a separate PR. 

In the worst case scenario of this Scrollbar usecase, when a user drags the scrollbar there will be a lag - it won't follow the mouse position.

In both cases either synchronous or asynchronous, if someone creates a custom implementation of something scrollable and their scroll implementation would suspend for some time/delay then the lag can be noticeable.

Reply:
> would suspend for some time/delay then the lag can be noticeable.

I haven't dived into suspention scrollables, but I heard that it uses different coroutine scope, that should perform scroll synchronously in one frame, just like in `buildSequence` + `yield`. But that just my speculation, it needs to be verified.

Reply:
The draft of async soulution is here: https://github.com/JetBrains/compose-multiplatform-core/pull/575
___

> but I heard that it uses different coroutine scope, that should perform scroll synchronously in one frame

In android tests and samples they explcitily use rememberCoroutineScope() to launch scrollTo. So there's no implicit/hidden special coroutine scope for scroll. Maybe I misundertood your point.
___

In k/js we can't block. But we can attempt to finish the scroll by the end of the frame (using Microtask dispatcher), unless a custom scroll implementation dispatches new coroutines using a different Dispatcher or delays - so coroutines can be dispatched in the future frames.

So it would really depend on the scroll implementation. With current implementations in compose it would work.

Reply:
> So there's no implicit/hidden special coroutine scope for scroll. Maybe I misundertood your point.

This was a guess, I don't have exact information about how it works, and what is the purpose of suspend scrollable.

Will try to dive more into it today/tomorrow.

Reply:
After looking at the code, Compose uses `LaunchedEffect` or `rememberCoroutineScope` to scroll Scrollable via usual touch gestures.

Also, taking into account the current ComposeScene mechanic, there won't be an input lag because of `ComposeScene.postponeInvalidate`. Scroll will be scheduled first, and the next frame only after it.

Let's change it to `launch` everywhere, not only in JS, because with `runBlocking` we could encounter a deadlock (`main thread -> sendPointerEvent -> runBlocking(MainDispatcher) -> custom scrollTo -> delay(10) -> wait main thread`)

Also, it is better to make code more cleaner. Instead of mutable properties rawPosition/position we should call a function in `onDragDelta`:
```
unscrolledDragDistance += dragDelta - sliderDelta
scope.launch {
    scrollToPosition(position + sliderDelta)
}
```

cc @m-sasha, will it be okay, if we will do these changes?

Reply:
P.S. In the future, maybe we should use `FlushDispatcher` for `LaunchedEffect`'s and call `flush` after each `sendPointerEvent`, to make Composable with effects predictable. But that is another thing (there are cons/pros)

Reply:
In this case we can proceed in 2 ways:

1. don't merge this PR, understand the issue more

2. use `runBlocking` in JVM/native and `launch` in web. Make Scrollabar as expect/actual (not `runblocking` itself) for marking it as Experimental on web, and create an issue to understand it later

Reply:
We can use `dispatchRawDelta` as we do with mouse wheel, but it won't work with nested scrolling. Maybe it nested scrolling issue though.

Reply:
I don't think scrollbars need nested scrolling. In fact, I think it would be a bug (dragging an inner scrollbar shouldn't scroll the outer scrollable). But I don't think that solves the problem. We still need snapToIndex, plus it's not clear what the situation is with scroll priority/cancellation.

By the way, it seems mouse-wheel actually does need nested scrolling to work, no?

Reply:
I think that unless we want to spend the time to understand the suspend scrolling better, the way to go is to do an expect/actual and launch it in a scope on JS.

Reply:
For now, I renamed the function to `runBlockingIfPossible` to make it more clear that it can be not blocking.



> If the value isn't updated before the next drag event is received, it will probably cause issues.
@m-sasha Sasha, there is draft which attempted to solve this:
https://github.com/JetBrains/compose-multiplatform-core/pull/575 - the values should get updated before next event is processed. WDYT?

Reply:
> I don't think scrollbars need nested scrolling

> it seems mouse-wheel actually does need nested scrolling to work, no?

I meant the case, when we use [nestedScrollConnection](https://medium.com/@debdut.saha.1/top-app-bar-animation-using-nestedscrollconnection-like-facebook-jetpack-compose-b446c109ee52)

The case, when we just have one more scroll inside another scroll is indeed isn't needed to be handled for scrollbars. This case also already works correctly with mouse (we scroll the nested scroll)